### PR TITLE
Initial work on `redis.cluster` module

### PR DESCRIPTION
**Pull request by sobolevn**
Source: https://github.com/redis/redis-py/blob/b2bcbc8bf1ae5622567acbc20e30e9c2e977e5c0/redis/cluster.py#L197

### CODE REVIEW

**Comment by sobolevn** [comment]
@AlexWaygood @srittau friendly ping 🙂

File: stubs/redis/redis/cluster.pyi [hunk 1]
```diff
@@ -1,75 +1,79 @@
...
     ERRORS_ALLOW_RETRY: ClassVar[tuple[type[RedisError], ...]]
-    user_on_connect_func: Any
-    encoder: Any
-    cluster_error_retry_attempts: Any
-    command_flags: Any
-    node_flags: Any
-    read_from_replicas: Any
+
+class RedisCluster(AbstractRedisCluster, RedisClusterCommands[_StrType], Generic[_StrType]):
+    user_on_connect_func: Callable[[Connection], object]
```
Comment (AlexWaygood) [inline stubs/redis/redis/cluster.pyi:44]: https://github.com/redis/redis-py/blob/c94821c11e74d360ae3ccdcf9581cfe24e120a07/redis/cluster.py#L606

```suggestion
    user_on_connect_func: Callable[[Connection], object] | None
```

File: stubs/redis/redis/cluster.pyi [hunk 2]
```diff
@@ -1,75 +1,79 @@
...
+    read_from_replicas: bool
     reinitialize_counter: int
     reinitialize_steps: int
-    nodes_manager: Any
-    cluster_response_callbacks: Any
-    result_callbacks: Any
-    commands_parser: Any
-    def __init__(
+    nodes_manager: NodesManager
+    cluster_response_callbacks: dict[str, Any]
```
Comment (AlexWaygood) [inline stubs/redis/redis/cluster.pyi:53]:
```suggestion
    cluster_response_callbacks: CaseInsensitiveDict[str, Callable[..., Incomplete]]
```

File: stubs/redis/redis/cluster.pyi [hunk 3]
```diff
@@ -1,75 +1,79 @@
...
+    result_callbacks: dict[str, Any]
```
Comment (AlexWaygood) [inline stubs/redis/redis/cluster.pyi:54]:
```suggestion
    result_callbacks: CaseInsensitiveDict[str, Callable[[Incomplete, Incomplete], Incomplete]]
```

File: stubs/redis/redis/cluster.pyi [hunk 4]
```diff
@@ -184,12 +188,14 @@ class ClusterPipeline(RedisCluster[_StrType], Generic[_StrType]):
...
+    def __exit__(
+        self, type: type[BaseException] | None, value: BaseException | None, traceback: TracebackType | None
+    ) -> None: ...
     def __del__(self) -> None: ...
-    def __len__(self): ...
-    def __nonzero__(self): ...
-    def __bool__(self): ...
+    def __len__(self) -> int: ...
+    def __nonzero__(self) -> bool: ...
+    def __bool__(self) -> bool: ...
```
Comment (AlexWaygood) [inline stubs/redis/redis/cluster.pyi:198]: These both always return `True`

```suggestion
    def __nonzero__(self) -> Literal[True]: ...
    def __bool__(self) -> Literal[True]: ...
```

File: stubs/redis/redis/cluster.pyi (hunk 3 above)
Comment (AlexWaygood) [reply stubs/redis/redis/cluster.pyi:55]: Why has this one been marked "resolved" but not been applied? :)

Comment (sobolevn) [reply stubs/redis/redis/cluster.pyi]: Now it is! 👍

Comment (AlexWaygood) [reply stubs/redis/redis/cluster.pyi]: Thanks!

File: stubs/redis/redis/cluster.pyi [hunk 5]
```diff
@@ -1,75 +1,80 @@
...
+from redis.client import CaseInsensitiveDict, PubSub, Redis
+from redis.commands import CommandsParser, RedisClusterCommands
 from redis.commands.core import _StrType
-from redis.connection import BaseParser, Encoder
+from redis.connection import BaseParser, Connection, Encoder
 from redis.exceptions import RedisError
 
-def get_node_name(host, port): ...
-def get_connection(redis_node, *args, **options): ...
+def get_node_name(host: str, port: int) -> str: ...
```
Comment (AlexWaygood) [inline stubs/redis/redis/cluster.pyi:14]: In the source code `port` is annotated as `Union[str, int]` -- any reason not to do that here? https://github.com/redis/redis-py/blob/c94821c11e74d360ae3ccdcf9581cfe24e120a07/redis/cluster.py#L41

```suggestion
def get_node_name(host: str, port: str | int) -> str: ...
```

**Review by AlexWaygood — APPROVED** [review]
Thanks!

**Comment by sobolevn** [comment]
Thank *you* for high-quality review!

**Pull request merged**